### ISSUE TITLE
8.6.8 breaks Timestamp support for Firestore Emulator

### ISSUE THREAD

**Issue by jirizavadil** [firebase/firebase-js-sdk#5047]
* Operating System version: MacOS 11.4
  * Browser version: Google Chrome 91.0.4472.106 (Official Build) (x86_64)
  * Firebase SDK version: 8.6.8
  * Firebase tools version: 9.13.1
  * Firebase Product: firestore

#### Steps to reproduce:
- update firebase to 8.6.8
- use firestore emulator
- write/read from firestore
- XHRs pending for 45s and then get cancelled

I've spent over 3 hours trying to replicate this in a new project, but unsuccessfuly. In the current project i could only try emulators, not cloud environment, as i only have production cloud env at this time. 

Last version that doesn't introduce any bugs for me is 8.4.0

**Comment by looptheloop88** [comment]
Hi @jirizavadil, thanks for the report. I tried replicating the issue, but I wasn't able to. I used the same SDK version and Firebase tools, but the test app was able to read and write data from Firestore database.

If I can replicate the issue, I can have a better look into it. Please share a [minimal, but complete](https://stackoverflow.com/help/minimal-reproducible-example) sample of a project that I can run locally. Also, you may enable Firestore debugging `firebase.firestore.setLogLevel("debug");` and share with us the logs.

**Comment by jsakas** [comment]
I spend most of my day yesterday trying to debug an issue which I believe might be related to this. In my case, using `snap.ref.update()` was updating the local db, but failing to update the emulated DB. Using `database.doc().update()` seemed to work fine. Spent hours trying to debug and create a repro unsuccessfully. Downgrading to v8.6.1 for now.

**Comment by mandalornl** [comment]
> I think I may have the same issue, but I've managed to isolate the problem. [...]

I would like to add that it's broken from `>= 8.6.3` and not related to `8.6.8` only.

Also notice the difference between the `timestampValue` values:
```json
Firebase <= 8.6.2

{
  "update": {
    "name": "projects/sometest-78e8d/databases/(default)/documents/test/CBVTnjvUzfraWgCuuynB",
    "fields": {
      "updatedAt": {
        "timestampValue": "2021-07-02T07:24:58.654000000Z"
      }
    }
  },
  "updateMask": {
    "fieldPaths": [
      "updatedAt"
    ]
  },
  "currentDocument": {
    "exists": true
  }
}
```

```json
Firebase >= 8.6.3

{
  "update": {
    "name": "projects/sometest-78e8d/databases/(default)/documents/test/eUikl3S7Bt6d4AxTq0aS",
    "fields": {
      "updatedAt": {
        "timestampValue": {
          "seconds": 1625210353,
          "nanos": 166000000
        }
      }
    }
  },
  "updateMask": {
    "fieldPaths": [
      "updatedAt"
    ]
  },
  "currentDocument": {
    "exists": true
  }
}
```

**Comment by jirizavadil** [comment]
@looptheloop88 here it is. Thanks to @mandalornl, i've been able to boil it down to a reproducible sample:

```
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Welcome to Firebase Hosting</title>

    <script defer src="/__/firebase/8.7.0/firebase-app.js"></script>
    <script defer src="/__/firebase/8.7.0/firebase-firestore.js"></script>
    <script defer src="/__/firebase/init.js?useEmulator=true"></script>
  </head>
  <body>
    <script>
      document.addEventListener("DOMContentLoaded", async function () {

        let db = firebase.firestore()
        db.useEmulator("localhost", 8432)

        var laRef = db.collection("cities").doc("LA")

        await laRef.set({
          created: new Date()
        })

        // WORKS
        await laRef.update({
          tralala1: true,
          updated1: firebase.firestore.FieldValue.serverTimestamp()
        })

        // DOES NOTHING
        await laRef.update({
          tralala2: true,
          updated2: new Date()
        })
      })
    </script>
  </body>
</html>
```

I've tested many versions with this result:
works:  8.4.0, 8.4.1, 8.4.2, 8.4.3, 8.5.0, 8.6.0, 8.6.1, 8.6.2
broken: 8.6.3, 8.6.4, 8.6.5, 8.6.7, 8.6.8, 8.7.0

**Comment by looptheloop88** [comment]
Hi @jirizavadil and @mandalornl, I was able to replicate the issue using the sample projects that you've shared with us. Let me bring this matter to the attention of our engineers here.

**Comment by schmidt-sebastian** [comment]
This broke with https://github.com/firebase/firebase-js-sdk/pull/4933, since the call [here](https://github.com/firebase/firebase-js-sdk/pull/4933/files#diff-733bdb5f107eab3c4508c06ccfc7f51e31d19610c95dc4886fe76c08c6059801R83) converts the timestamp to JSON (it should be a string).

Will be fixed next week.

**Comment by silviokucz** [comment]
This issue is still happening. When I try changing the data in a timestamp field, it seems like the change took but when I refresh the page it still shows the old timestamp. No entries in the log!
I have firebase version 9.16.0

**Comment by schmidt-sebastian** [comment]
Do you have persistence enabled? I wonder if these documents are read from cache.

You should be able to see logs from the Firestore SDK itself (call `firebase.firestore.setLogLevel('debug')`).

**Comment by silviokucz** [comment]
This is how I start the emulator:
"npm run build:watch && firebase emulators:start --inspect-functions --export-on-exit c:\\tmp\\firebase --import c:\\tmp\\firebase"

Is there a way to set the log level using the CLI?

**Comment by schmidt-sebastian** [comment]
No - this would be a line you need to add to your Web or Node app.

**Comment by silviokucz** [comment]
Well, I'm not sure how to do that on the functions side, is there an example?
where do i import firebase from?

I have this but doesn't compile:.
import * as functions from 'firebase-functions'
...
functions.firestore.setLogLevel('debug')

(should be easy for you to try on your end, do you experience the same problem?)

**Comment by silviokucz** [comment]
Update:
I was able to set debug level using this flag: 
firebase **--debug**  emulators:start --inspect-functions --export-on-exit c:\\tmp\\firebase --import c:\\tmp\\firebase
I see a lot of logs starting with [debug]
Unfortunately working with the firestore emulator and editing values doesn't generate any entries.

**Comment by schmidt-sebastian** [comment]
Are you writing to Firestore from your function? You may have to invoke the Admin SDK directly. If you can send me a code snippet, I can modify it to enable logging.

**Comment by mandalornl** [comment]
> This issue is still happening. Using the emulator, when I try changing the data in a timestamp field, it seems like the change took but when I refresh the page it still shows the old timestamp. No entries in the log!
> I have firebase version 9.16.0

I want to confirm that this happens when you try to add or update a `Timestamp` field from an existing `Document` using the `Firebase Emulator Suite` web interface on `localhost:4000`. This also happend before the update in versions `> 8.6.2` and `< 8.8.0`.

**Comment by silviokucz** [comment]
Confirmed. Using the emulator served on localhost:4000. Changing the value of a timestamp field fails silently.

**Comment by schmidt-sebastian** [comment]
I was able to write Timestamps from the Web SDK, the Node Web SDK and the Node Admin SDK successfully.

Here is my setup: https://gist.github.com/schmidt-sebastian/59c76a6655b3556d542b90882de689bc

Can you spot a difference?

**Comment by silviokucz** [comment]
Thanks for sharing the code. Writing and reading timestamp fields using the SDK is not the issue, maybe I didn't explain well on the original post.
The issue occurs when using the emulator web UI manually.

**Comment by schmidt-sebastian** [comment]
Ah, okay, that makes sense. Will get this fixed.

**Comment by yuchenshi** [comment]
Fixed in Emulator UI in Firebase CLI v9.16.1.
